qenspy/qens.py

```diff
@@ -194,10 +273,151 @@ def predict(self, q, w):
+        w: 2D tensor with shape (K, M)
+            Model weight. Not required if self.bw_method is “unweighted”
+        rectangle_bw: 2D tensor with shape (K, M)
+            Rectangle bandwidth values
+
+        Returns
+        -------
+        weighted_cdf: 3D tensor with shape (N, K, ...)
+            weighted_cdf
+        """
+
+        # to handle missing values
+        # (N, K, M)
+        broadcast_w = super().broadcast_w_and_renormalize(q, w)
+
+        #(N, K, ...)
+        weighted_cdf = tf.zeros(shape = x.shape, dtype = x.dtype)
+        M = q.shape[2]
+        for i in range(M):
+            # (N, K, 1)
+            low = tf.expand_dims(tf.subtract(q[:,:,i], rectangle_bw/2), -1)
+            high = tf.expand_dims(tf.add(q[:,:,i], rectangle_bw/2), -1)
+            curr_broadcast_w = tf.expand_dims(broadcast_w[:,:,i],-1)
+            # case1 (no calculation needed): when x is on the left hand side of the rectangular kernel
+            # case2: when x is in the middle of the rectangular kernel
+            weighted_cdf = tf.where(tf.logical_and(tf.less_equal(x, high), \
+                                tf.greater_equal(x, low)), \
+                                tf.add(weighted_cdf, curr_broadcast_w * tf.subtract(x, low) * (1/  tf.expand_dims(rectangle_bw,-1))),\
+                                weighted_cdf)
+            # case3: when x is on the right hand side of the rectangular kernel
+            weighted_cdf = tf.where(tf.greater(x, high), tf.add(weighted_cdf, curr_broadcast_w), weighted_cdf)
+
+        return weighted_cdf
+    
+    def predict(self, q, w):
+        """
+        Calculate weighted median at different quantile levels
+
+        Parameters
+        ----------
+        q: 3D tensor with shape (N, K, M)
+            Predictive quantiles from component models where N is the number of
+            location/forecast date/horizon triples, M is number of models,
+            and K is number of quantile levels
+        w: 2D tensor with shape (K, M)
+            Model weight. Not required if self.bw_method is “unweighted”
+        
+        Returns
+        -------
+        median: 3D tensor with shape (N, K)
+            weighted median
+        """
+
+        bw = self.calc_bandwidth(q = q, w = w)
+        rectangle_bw = tf.sqrt(12 * (bw ** 2))
+
+        low = q - tf.reshape(rectangle_bw/2, [rectangle_bw.shape[0], rectangle_bw.shape[1], 1])
+        high = q + tf.reshape(rectangle_bw/2, [rectangle_bw.shape[0], rectangle_bw.shape[1], 1])
+        # (N, K, 2M)
+        slope_changepoints = tf.sort(tf.concat([low, high], axis = 2))
+
+        # (N, K, 2M)
+        changepoint_cdf_values = self.weighted_cdf(x = slope_changepoints, q = q, w = w, rectangle_bw = rectangle_bw)
+        
+        # the smallest index that has cdf >= 0.5 (N, K)
+        inds = tf.math.argmax(tf.where(changepoint_cdf_values >= np.float64(0.5), np.float64(0.5)-changepoint_cdf_values, np.float64(-1.0)), axis = 2)
```

Comment:
Can we simplify by dropping the where?
```suggestion
        inds = tf.math.argmax(changepoint_cdf_values >= np.float64(0.5), axis = 2)
```